range_tombstone_list.hh

```diff
@@ -98,10 +98,14 @@ class range_tombstone_list final {
     };
     class update_undo_op {
         range_tombstone _old_rt;
-        const range_tombstone_entry& _new_rt;
+        range_tombstone_entry& _new_rt;
     public:
-        update_undo_op(range_tombstone&& old_rt, const range_tombstone_entry& new_rt)
-                : _old_rt(std::move(old_rt)), _new_rt(new_rt) { }
+        update_undo_op(range_tombstone& new_rt, range_tombstone_entry& new_rt_entry) noexcept
+            : _old_rt(std::move(new_rt_entry.tombstone()))
+            , _new_rt(new_rt_entry)
+        {
+            new_rt_entry.tombstone() = std::move(new_rt);
```

Comment:
If I understand correctly, this is to replace the `swap` which you previously did in `update`.

I wonder if it's right that the undoer also does the update now. Philosophically, undoer should only be responsible for undoing, not doing the update itself. The code now is also inconsistent with `erase`, where the erase is done inline in `erase` and not in the undoer. While for `update`, the update is done in the undoer.

Couldn't you do the `swap` after the `emplace_back`?

Alternatively, we could also modify `erase` and `insert` to do the operation inside the undoer - it's actually consistent with RAII philosophy, constructor does something, destructor undoes it - but then the undoer would IMO have to be renamed, because it's now a wrapper both for the update and the undo of the update.

Reply:
I see your point. I kept erase_undo_op unchanged in V2.